execution/aggregate/khashaggregate.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/exp/slices"
+
+	"github.com/thanos-community/promql-engine/execution/model"
+)
+
+type kAggregate struct {
+	next    model.VectorOperator
+	paramOp model.VectorOperator
+
+	vectorPool *model.VectorPool
+
+	by          bool
+	labels      []string
+	aggregation parser.ItemType
+
+	once        sync.Once
+	series      []labels.Labels
+	inputToHeap []*samplesHeap
+	heaps       []*samplesHeap
+	compare     func(float64, float64) bool
+}
+
+func NewKHashAggregate(
+	points *model.VectorPool,
+	next model.VectorOperator,
+	paramOp model.VectorOperator,
+	aggregation parser.ItemType,
+	by bool,
+	labels []string,
+) (model.VectorOperator, error) {
+	var compare func(float64, float64) bool
+
+	if aggregation == parser.TOPK {
+		compare = func(f float64, s float64) bool {
+			return f < s
+		}
+	} else {
+		compare = func(f float64, s float64) bool {
+			return s < f
+		}
+	}
+	// Grouping labels need to be sorted in order for metric hashing to work.
+	// https://github.com/prometheus/prometheus/blob/8ed39fdab1ead382a354e45ded999eb3610f8d5f/model/labels/labels.go#L162-L181
+	slices.Sort(labels)
+
+	a := &kAggregate{
+		next:        next,
+		vectorPool:  points,
+		by:          by,
+		aggregation: aggregation,
+		labels:      labels,
+		paramOp:     paramOp,
+		compare:     compare,
+	}
+
+	return a, nil
+}
+
+func (a *kAggregate) Next(ctx context.Context) ([]model.StepVector, error) {
+	in, err := a.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		return nil, nil
+	}
+
+	defer a.next.GetPool().PutVectors(in)
+
+	var args []model.StepVector
+
+	if a.paramOp != nil {
+		args, err = a.paramOp.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if len(args) != len(in) {
+		return nil, fmt.Errorf("scalar argument not found")
+	}
+
+	a.once.Do(func() { err = a.init(ctx) })
+	if err != nil {
+		return nil, err
+	}
+
+	result := a.vectorPool.GetVectorBatch()
+	for i, vector := range in {
+		a.aggregate(vector.T, &result, int(args[i].Samples[0]), vector.SampleIDs, vector.Samples)
+		a.next.GetPool().PutStepVector(vector)
+	}
+
+	return result, nil
+}
+
+func (a *kAggregate) Series(ctx context.Context) ([]labels.Labels, error) {
+	var err error
+	a.once.Do(func() { err = a.init(ctx) })
+	if err != nil {
+		return nil, err
+	}
+
+	return a.series, nil
+}
+
+func (a *kAggregate) GetPool() *model.VectorPool {
+	return a.vectorPool
+}
+
+func (a *kAggregate) Explain() (me string, next []model.VectorOperator) {
+	if a.by {
+		return fmt.Sprintf("[*kaggregate] %v by (%v)", a.aggregation.String(), a.labels), []model.VectorOperator{a.paramOp, a.next}
+	}
+	return fmt.Sprintf("[*kaggregate] %v without (%v)", a.aggregation.String(), a.labels), []model.VectorOperator{a.paramOp, a.next}
+}
+
+func (a *kAggregate) init(ctx context.Context) error {
+	series, err := a.next.Series(ctx)
+	if err != nil {
+		return err
+	}
+	hapsHash := make(map[uint64]*samplesHeap)
```

Comment:
What does this map hold, and can we make it a slice instead? So far we tried to avoid using maps for lookups because they tend to slow things down. This is why we assign a zero-based ID to each series and we use that as a key into a slice.

Reply:
This has is only to calculate the `by` `without` aggragation.  In other words, basically a need one heap per `by` value.

This is similar to what is done on:

https://github.com/thanos-community/promql-engine/blob/f0c54531a5ee47fcc80d57f97cbb362914b02f9b/execution/aggregate/hashaggregate.go#L185-L194

Reply:
Got it. I think this can be optimize into a slice. I will look into it as a follow up.